### PR TITLE
Professor application form

### LINKED ISSUE #660: Implement professor application form

### PR DESCRIPTION
**Pull request by liustanley**
# Description

Adds a new page for an application form for professors to apply to use our app.
Submitted courses via this application form create `CourseModels` and `CourseSectionMappingModels` that aren't enabled until approved (to be done in next PR).

Only hairy part of this PR is [this commit](https://github.com/sandboxnu/office-hours/commit/2b9e4e95b660e16a7bdc54ae769ad0d9e8c2f187), I refactored some code to filter out courses that have `enabled=false`, since that's how we create submitted courses. Please look at this part carefully!

TODO:
- [x] Add password to protect form URL
- [x] Migrations
- [ ] Maybe write tests

Closes #660 

## Type of change

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

So far only manually tested to verify form creates new models. In order to test manually, you need to create a new `Semester` locally for `season=Summer_1, year=2021`.

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code where needed 
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/app/components/Apply/ApplyPage.tsx [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            Submit another course
+          </Button>,
+        ]}
+      />
+    );
+  }
+
+  return (
+    <div style={{ padding: "3% 12%" }}>
+      <h1>Apply for Khoury Office Hours</h1>
```
Comment (willstenzel) [inline packages/app/components/Apply/ApplyPage.tsx:70]:
```suggestion
      <h1>New Course Application for Khoury Office Hours</h1>
```

File: packages/app/components/Apply/ApplyPage.tsx [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+          <HalfFormItem
+            label="Semester"
+            name="semester"
+            rules={[{ required: true, message: "Please select a semester." }]}
+          >
+            <Select defaultValue="Summer_1 2021">
+              <Option value="Summer_1 2021">Summer 1 2021</Option>
+              <Option value="Summer_2 2021">Summer 2 2021</Option>
+              <Option value="Summer_Full 2021">Summer Full 2021</Option>
+            </Select>
```
Comment (willstenzel) [inline packages/app/components/Apply/ApplyPage.tsx:137]: Could we pull these from our semester model?

**Review by willstenzel — APPROVED** [review]
Looks good! 

I took a look at https://github.com/sandboxnu/office-hours/commit/2b9e4e95b660e16a7bdc54ae769ad0d9e8c2f187 and I didn't see any issues but it might be worth writing a test for methods in the login service like `courseSectionToCourse` since we aren't able to test that by running the app locally. I'd be happy to help with that later today.

File: packages/server/src/course/course.controller.ts [hunk 3]
```diff
@@ -301,6 +304,40 @@ export class CourseController {
...
+      semesterId: semester.id,
+      enabled: false,
+      pending: true,
+      timezone: body.timezone,
+    }).save();
+
+    // create CourseSectionMappings for each submitted section number
+    body.sections.forEach(async (section) => {
+      await CourseSectionMappingModel.create({
+        genericCourseName: body.name,
```
Comment (IrisLiu-00) [inline packages/server/src/course/course.controller.ts:334]: Do we care at all about normalizing course names from the form input (eg. cs3000 vs CS 3000 vs Algorithms and Data)? It seems like the user-entered course name gets also saved as this `genericCourseName` - which is supposed to match with what Khoury Admin gives, right?

**Review by IrisLiu-00 — APPROVED** [review]

File: packages/app/components/Apply/ApplyPage.tsx [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+  };
+
+  const resubmit = () => {
+    form.resetFields();
+    setSubmitted(false);
+  };
+
+  const validateSections = (value: string): boolean => {
+    const values = value.replace(/\s+/g, "").split(",");
+    for (const val of values) {
```
Comment (NEUDitao) [inline packages/app/components/Apply/ApplyPage.tsx:39]: https://docs.racket-lang.org/reference/pairs.html#%28def._%28%28lib._racket%2Fprivate%2Fmap..rkt%29._andmap%29%29

File: packages/app/components/Apply/ApplyPage.tsx [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    const split = value.split(" ");
+    const subject =
+      split[0] == "CS" ||
+      split[0] == "DS" ||
+      split[0] == "CY" ||
+      split[0] == "IS";
+    return split.length == 2 && subject && /^\d+$/.test(split[1]);
+  };
+
+  if (router.query[process.env.APPLY_KEY] !== "") {
```
Comment (NEUDitao) [inline packages/app/components/Apply/ApplyPage.tsx:55]: don't we wanna CSRF it up here?

File: packages/app/components/Apply/ApplyPage.tsx [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        Please submit an application for each course you want to use Khoury
+        Office Hours for.
+      </div>
+      <br />
+      <Form
+        form={form}
+        layout="vertical"
+        initialValues={{
+          remember: true,
+          semester: "Summer_1 2021",
```
Comment (NEUDitao) [inline packages/app/components/Apply/ApplyPage.tsx:103]: wonder if this should be an env variable like "curr semester"

**Review by NEUDitao — APPROVED** [review]
image

File: packages/app/components/Apply/ApplyPage.tsx (hunk 5 above)
Comment (liustanley) [reply packages/app/components/Apply/ApplyPage.tsx:55]: i still have no idea what u mean

**Pull request merged**